### DISCUSSION TITLE
Outputing a REAL - not cm3 specific

### DISCUSSION THREAD

**Discussion by dukester** [modula3/cm3#1162] [General]
Although the issue that I bring here has to do with Oberon-2 code, it could very easily be cm3 code. There seems to be a few mathematicians here so I thought I'd take the liberty and ask for directions here.

I wrote a simple temperature conversion utility - menu driven, C to F or F to C.
It works fine, but in Oberon-2, REALs are output in scientific notation only.

```
Temperature Conversion Utility
------------------------------
1 - Fahrenheit TO Celsius
2 - Celsius TO Fahrenheit
  - Enter your Choice
1
Enter a number/temperature TO convert: 32

Your input was 3.200000E+01 Fahrenheit!
That's 0.000000E+00 In Celsius!
```
Would any of you know of a way to convert the above output to a one decimal place REAL?
My goal is to port the code cm3 as a learning exercise, but I want to get my head around of how to tweak the above output to something a non math guru would understand. TIA!

**Comment by RodneyBates** [comment]
Have a look at the Fmt interface in m3-libs/libm3/src/fmtlex/Fmt.i3.
Procedure Real formats a REAL into characters in several alternative
ways,  I would guess Style.Fix, with prec as the desired number of
fractional digits, would probably be what you would want.

**Reply by dukester** [reply]
Thx Rodney! I'll sure give your Reference a good study. God gave me a lot of talents, but a mathematical mind isn't one of them. I could never "see" the point of calculus etc - until I decided to build my own bamboo flyrod. Everett Garrison, a structural engineer, wrote a book on building bamboo flyrods. He got into the modulus of elasticity of bamboo, taper derivations - the whole nine yards. It wasn't until he explained about getting taper dimensions by calculating - long-hand - the area under the curve created by a flexed flyrod, that I understood what calculus was trying to do. Throwing the theory at me didn't do it. Applying the theory to something I enjoyed and could relate to did the trick. 

Thanks again for the reference!

**Reply by torstenk** [reply]
For new and casual Modula-3 users coming across this as a search result:

[Fmt.i3](https://www.cs.purdue.edu/homes/hosking/m3/help/gen_html/libm3/src/fmtlex/Fmt.i3.html)

[Quick ref](https://cs.fit.edu/~ryan/m3/tutorial/html/m3.html)

```
MODULE Main;

IMPORT IO, Fmt; 
FROM Fmt IMPORT Style;

VAR fahrenheit: REAL; 

PROCEDURE FtoC(fahrenheit: REAL): REAL =
  BEGIN
    RETURN FLOAT(5) * (fahrenheit - 32.0) / FLOAT(9)
  END FtoC;

BEGIN
  IO.Put("Enter the temperature in Fahrenheit: ");
  TRY
    fahrenheit := IO.GetReal();
    IO.Put(   Fmt.Real(fahrenheit,       style := Style.Fix, prec := 1) & " F = " 
            & Fmt.Real(FtoC(fahrenheit), style := Style.Fix, prec := 1) & " C\n");
  EXCEPT IO.Error => IO.Put("Cannot read that as a floating point number, aborting.\n")
  END
END Main.
```

It's been a while...
